### PR TITLE
Sgdemux

### PR DESCRIPTION
**Pull request by sam-white04**
Opening a draft PR for increased readability of changes to `nf-core-modules/modules/nf-core/sgdemux/` & `nf-core-modules/tests/modules/nf-core/sgdemux`

### CODE REVIEW

File: modules/nf-core/sgdemux/main.nf [hunk 1]
```diff
@@ -0,0 +1,49 @@
+process SGDEMUX {
+    tag "$meta.id"
+    label 'process_high'
+
+    conda "bioconda::sgdemux=1.1.0"
+    if (workflow.profile.tokenize(',').intersect(['Singularity', 'Docker']).size() >= 1) {
+        exit 1, "SGDEMUX module does not support Singularity or Docker. Please use Conda / Mamba instead."
+    }
+    // TODO check sgdemux is not on singularity/docker (google says no) double check with Nils
```
Comment (nh13) [inline modules/nf-core/sgdemux/main.nf]: Having `sgdemux` in `bioconda` will make a docker container accessible on `quay.io`.  See https://bioconda.github.io/recipes/sgdemux/README.html#package-sgdemux, in particular:
`quay.io/biocontainers/sgdemux:<tag>`

For singularity, we'll need a `Dockerfile` ([see requirement (5)](https://nf-co.re/developers/modules#software-requirements)).  Feel free to ask the team how to create a dockerfile (or look for examples in other modules in nf-core/modules)

File: modules/nf-core/sgdemux/main.nf [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    // container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ? 
+        // 'https://depot.galaxyproject.org/singularity/YOUR-TOOL-HERE':
+        // 'quay.io/biocontainers/YOUR-TOOL-HERE' }"]
+
+    input:
+    tuple val(meta), path(run_manifest), path(run_dir)
```
Comment (nh13) [inline modules/nf-core/sgdemux/main.nf]:
```suggestion
    tuple val(meta), path(sample_sheet), path(in_fastqs_dir)
```

File: modules/nf-core/sgdemux/main.nf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    output:
+    // TODO output/Samples/*/*_R*...
```
Comment (nh13) [inline modules/nf-core/sgdemux/main.nf]:
```suggestion
```

File: modules/nf-core/sgdemux/main.nf (hunk 2 above)
Comment (nh13) [reply modules/nf-core/sgdemux/main.nf]: I'd also add a comment either here or above the process that the input FASTQs must be bgzipped!

File: modules/nf-core/sgdemux/main.nf [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    tuple val(meta), path('output/*_R*.fastq.gz'), emit: sample_fastq
+    tuple val(meta), path('output/metrics.tsv'), emit: metrics
+    tuple val(meta), path('output/most_frequent_unmatched.tsv'), emit: most_frequent_unmatched
+    tuple val(meta), path('output/per_project_metrics.tsv'), emit: per_project_metrics
+    tuple val(meta), path('output/per_sample_metrics.tsv'), emit: per_sample_metrics
+    tuple val(meta), path('output/sample_barcode_hop_metrics.tsv'), emit: sample_barcode_hop_metrics
+    path "versions.yml"           , emit: versions
```
Comment (nh13) [inline modules/nf-core/sgdemux/main.nf]: align emits

File: modules/nf-core/sgdemux/main.nf [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def runManifest = run_manifest ? "-r ${run_manifest}" : ""
```
Comment (nh13) [inline modules/nf-core/sgdemux/main.nf]:
```suggestion
    def sampleSheet = run_manifest ? "-r ${run_manifest}" : ""
```

File: modules/nf-core/sgdemux/main.nf [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    """
+    mkdir -p output/
+    sgdemux \\
+        --sample-metadata $run_manifest \\
+        --fastqs $run_dir \\
+        --output-dir output/ \\
+        --demux-threads $task.cpus \\
+        --compressor-threads $task.cpus \\
+        --writer-threads $task.cpus \\
```
Comment (nh13) [inline modules/nf-core/sgdemux/main.nf]:
```suggestion
        --writer-threads $task.cpus
```

File: modules/nf-core/sgdemux/meta.yml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      documentation: "https://github.com/Singular-Genomics/singular-demux"
+      licence: "['For Singular G4™ Sequencing Platform only']"
+
+input:
+- meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - run_manifest:
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml]: Can you use sample sheet here and in the descriptionb elow

File: modules/nf-core/sgdemux/meta.yml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      type: file
+      description: RunManifest file
+      pattern: "*.{csv}"
+  - run_dir:
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml]:
```suggestion
  - fastqs_dir:
```

File: modules/nf-core/sgdemux/meta.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      type: directory
+      description: "Input run directory containing optionally containing a RunManifest.json if run_manifest is not supplied"
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml]: Please update based on docs.  This should be a directory containing the FASTQs.

File: modules/nf-core/sgdemux/meta.yml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - sample_fastq:
+      type: file
+      description: Demultiplexed sample FASTQ files
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml]:
```suggestion
      description: Demultiplexed per-sample FASTQ files
```

File: modules/nf-core/sgdemux/meta.yml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+      pattern: "output/*_R*.fastq.gz"
+  - metrics:
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml:41]: update here and below based on https://github.com/Singular-Genomics/singular-demux#metrics

File: modules/nf-core/sgdemux/nextflow.config [hunk 12]
```diff
@@ -0,0 +1,46 @@
+params {
```
Comment (nh13) [inline modules/nf-core/sgdemux/nextflow.config]: I don't think we need this file.  I don't see it in `fastqc` or `bcl2fastq` for example.  Thoughts?

File: tests/modules/nf-core/sgdemux/main.nf [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+nextflow.enable.dsl = 2
+
+include { SGDEMUX } from '../../../../modules/nf-core/sgdemux/main.nf'
+
+workflow test_sgdemux {
+    
+    // input = Channel.value([
+    //    [ id: 'sim-data' ], // meta map
+    //    file("/Users/.../projects/SINGULAR/singular-demux/simulated/out.sample_meta.csv", checkIfExists: true),
+    //    file("/Users/.../projects/SINGULAR/singular-demux/simulated/", checkIfExists: true)
```
Comment (nh13) [inline tests/modules/nf-core/sgdemux/main.nf]: TODO: add a tarball to `nf-core/test-datasets`, then have a process to untar it prior to running SGDEMUX, similar to: https://github.com/nf-core/modules/blob/master/tests/modules/nf-core/bases2fastq/main.nf

File: tests/modules/nf-core/sgdemux/nextflow.config [hunk 14]
```diff
@@ -0,0 +1,5 @@
+process {
+
+    publishDir = { "${params.outdir}/${task.process.tokenize(':')[-1].tokenize('_')[0].toLowerCase()}" }
+
```
Comment (nh13) [inline tests/modules/nf-core/sgdemux/nextflow.config:8]: as per bases2fastq 
```suggestion
     withName: UNTAR {
        publishDir = [ enabled: false ]
    }
```

File: tests/modules/nf-core/sgdemux/test.yml [hunk 15]
```diff
@@ -0,0 +1,11 @@
+## TODO nf-core: Please run the following command to build this file:
+#                nf-core modules create-test-yml sgdemux
```
Comment (nh13) [inline tests/modules/nf-core/sgdemux/test.yml]:
```suggestion
```

File: modules/nf-core/sgdemux/main.nf (hunk 1 above)
Comment (nh13) [reply modules/nf-core/sgdemux/main.nf]: Feel free to ping the nf-core folks on slack for help on this too!

Comment (sam-white04) [reply modules/nf-core/sgdemux/main.nf]: I see `sgdemux` on `quay.io`, and I will make changes to line 12 in the next commit to reflect that. I am happy to create a `Dockerfile` for `sgdemux`. I did not know that was up to us

File: modules/nf-core/sgdemux/main.nf (hunk 2 above)
Comment (sam-white04) [reply modules/nf-core/sgdemux/main.nf]: Agreed, in fact I already added that locally before seeing this.  Should we add some type of check to ensure fastq files are in the correct format?

File: modules/nf-core/sgdemux/nextflow.config (hunk 12 above)
Comment (sam-white04) [reply modules/nf-core/sgdemux/nextflow.config]: You are correct, `fastqc` nor `bcl2fastq` have `nextflow.config` files I will remove this file for sgdemux.

File: modules/nf-core/sgdemux/main.nf (hunk 1 above)
Comment (sam-white04) [reply modules/nf-core/sgdemux/main.nf]: @nils I am a bit confused about these container requirements. Requirement 5 in the link you provided makes me think that a `Dockerfile` must be provided only if software ifsnot available on `bioconda`. So if sgdemux is already available via `bioconda` and  `quay.io`, do we need `singularity`?

**Comment by sam-white04** [comment]
Here is the latest version of [sgdemux](https://github.com/fulcrumgenomics/nf-core-modules/commit/be297803c88936c8a665fa52e80199a1234371a8), including the checks shown below. Most importantly `test.yml` has been updated to exclude md5sums for fastq's fue to sorting errors. 

## PR checklist

- [x] This comment contains a description of changes (with reason).

- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

File: modules/nf-core/sgdemux/meta.yml [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+description: Demultiplex bgzip'd fastq files
+keywords:
+  - demultiplex
+  - fastq
+tools:
+  - "sgdemux":
+      description: "Tool for demultiplexing sequencing data generated on
+      Singular Genomics' sequencing instruments."
+      homepage: "https://github.com/Singular-Genomics/singular-demux"
+      documentation: "https://github.com/Singular-Genomics/singular-demux#per_project_metricstsv"
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml]: Top of readme?
```suggestion
      documentation: "https://github.com/Singular-Genomics/singular-demux#sgdemux"
```

File: modules/nf-core/sgdemux/meta.yml [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - sample_sheet:
+      type: file
+      description: sample_sheet file (two column csv wih Sample_Barcode and Sample_ID)
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml]:
```suggestion
      description: sample_sheet file (either a Singular Genomics sample sheet, or a two column csv wih Sample_Barcode and Sample_ID)
```

File: modules/nf-core/sgdemux/meta.yml [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+      pattern: "*.{csv}"
+  - fastqs_dir:
+      type: directory
+      description: "Input directory containing bgzipped FASTQ files"
```
Comment (nh13) [inline modules/nf-core/sgdemux/meta.yml]:
```suggestion
      description: "Input directory containing bgzipped (not gzip) FASTQ files"
```

**Review by nh13 — APPROVED** [review]
Minor comments, looks good.  Fixup and then retarget to the main nf-core/modules repo!

**Pull request closed without merging**